### PR TITLE
Feature/hxgkdw8s add the todo table and model

### PR DESCRIPTION
**Pull request by hannahbit**
My first time setting the associations of parent(list) and child(todo) up...
I especially struggled with making sure that todos are only saved when the are associated with an existing list.
Any feedback is very welcome :) 

Also, I know I should really start working with factories on the tests. Will do so soon!

### CODE REVIEW

File: lib/do_it/todo.ex [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    field :description, :string
+
+    timestamps()
+  end
+
+  @doc false
+  def create_changeset(todo, attrs) do
+    todo
+    |> cast(attrs, [:description, :checked])
+    |> validate_required([:description, :list_id])
```
Comment (hannahbit) [inline lib/do_it/todo.ex:17]: Is this ok? Putting the `:list_id` as required? In the documentation is written that you should use `cast_assoc` with option `required: true` instead. But it also seemed to me that you use `cast_assoc` only on the parent. If I didn't put the `list_id` here, todos without an associated list would be saved.

File: lib/do_it/todo.ex [hunk 2]
```diff
@@ -0,0 +1,21 @@
+defmodule DoIt.Todo do
+  use Ecto.Schema
+  import Ecto.Changeset
+
+  schema "todos" do
+    belongs_to :list, DoIt.List
+    field :checked, :boolean, default: false
+    field :description, :string
```
Comment (tessi) [inline lib/do_it/todo.ex:8]: this schema sounds like a good way to store todo items 👍 

hmmm, this is probably just an opinion, but I would personally rename the `checked` field to something like `done` (depending on what you mean with an item being "checked"). A more specific name can open the door for future features (maybe you want two checks, one for "done", one for "approved" or so 🤔 )

File: lib/do_it/todo.ex (hunk 1 above)
Comment (tessi) [reply lib/do_it/todo.ex:17]: good question, I often pass a required external struct as an explicit param, e.g. (and I just typed that code in a text field here, maybe it doesn't compile 🙈 )

```
def create_changeset(todo, attrs, %DoIt.List{} = list) do
  todo
  |> cast([:description, :checked])
  |> validate_required([:description])
  |> put_assoc(:list, list)
```

This way, you have no choice but to pass in an actual list.

File: priv/repo/migrations/20210610151713_create_todos.exs [hunk 3]
```diff
@@ -0,0 +1,15 @@
+defmodule DoIt.Repo.Migrations.CreateTodos do
+  use Ecto.Migration
+
+  def change do
+    create table(:todos) do
+      add :description, :string
```
Comment (tessi) [inline priv/repo/migrations/20210610151713_create_todos.exs:6]: hmm, do you want descriptions be null in the db? it looks like you have a test assuming it can't be nil 🤔

File: lib/do_it/todo.ex [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+    timestamps()
+  end
+
+  @doc false
+  def create_changeset(todo, attrs) do
+    todo
+    |> cast(attrs, [:description, :checked])
+    |> validate_required([:description, :list_id])
+    |> validate_length(:description, min: 3, max: 100)
```
Comment (tessi) [inline lib/do_it/todo.ex:18]: uh, does this work well with the description allowed to be `nil`? 💡  didn't know

File: test/do_it/todo_test.exs [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  end
+
+  test "todo associated with existing list and description
+        with 3-100 chars is saved to the database" do
+    {:ok, list} = Repo.create_list(%{title: "New List"})
+    todo = Ecto.build_assoc(list, :todos)
+    todo_changeset = Todo.create_changeset(todo, %{description: "Water plants"})
+    assert {:ok, todo} = Repo.insert(todo_changeset)
+    assert todo.list_id == list.id
+  end
```
Comment (tessi) [inline test/do_it/todo_test.exs:47]: I think this is a great test to get to know the ecto API and play around with todo items, but I think this shouldn't land in the final code. It doesn't add much value on top of your (great) tests above as it just tests Ecto API.

Probably (in a later stage) you will have similar code in a controller or access/service module - if that is the case, that code can and should be tested :)

File: test/do_it/todo_test.exs [hunk 6]
```diff
@@ -0,0 +1,48 @@
+defmodule DoIt.TodoTest do
+  use DoIt.DataCase, async: true
+  alias DoIt.Todo
+  alias DoIt.Repo
+
+  @long_description "This is a very long description. It is longer than onehundred characters. This is really very long. Who would give their todo list such a long description?"
```
Comment (tessi) [inline test/do_it/todo_test.exs:6]: hmm, I like that this is a very explicit way of demonstrating a very long description. But it's hard to verify with ones eyes that it's really more than 100 characters. Maybe the test becomes easier to read if you generate the String.

maybe using 

```
String.duplicate("a", 101)
```

you can read up on `duplicate/2` or even find a better method in the official string API 👉  https://hexdocs.pm/elixir/1.12/String.html

File: test/do_it/todo_test.exs [hunk 7]
```diff
@@ -0,0 +1,48 @@
+defmodule DoIt.TodoTest do
```
Comment (tessi) [inline test/do_it/todo_test.exs:1]: I find the tests read quite well 🌟 

one thing you could add (for future maintainability, not *really* for the size of your current codebase) is grouping the tests by the method you test.

That can be done using the `describe` macro like:

```
describe "changeset/2" do
  test "a good description of the edge case" do
    # ...
  end

  test "a good description of some failure case" do
    # ...
  end

end
```

This *really* helps organizing tests once you have more than one method to test.

**Review by tessi — APPROVED** [review]
Looks real clean and well done 💪 . I have some suggestions here and there, some based on what I _think_ you want without actually knowing :) Feel very much welcome to discuss or ask.

File: lib/do_it/todo.ex (hunk 1 above)
Comment (hannahbit) [reply lib/do_it/todo.ex:17]: That's a nice idea! I was so focused on the pipes that I didn't think about that 😅

File: lib/do_it/todo.ex (hunk 2 above)
Comment (hannahbit) [reply lib/do_it/todo.ex:8]: Ha. Good point!

File: priv/repo/migrations/20210610151713_create_todos.exs (hunk 3 above)
Comment (hannahbit) [reply priv/repo/migrations/20210610151713_create_todos.exs:6]: Ah. No, I don't want that. 🙅‍♀️

File: lib/do_it/todo.ex (hunk 4 above)
Comment (hannahbit) [reply lib/do_it/todo.ex:18]: I don't fully understand... I thought having a validation AND a database restriction is just to make double sure, that no unwanted data is stored. How would the database now, what the validations are doing?
Or do you mean the case, that I smuggle some todo with a nil description into the DB without validating before?

File: test/do_it/todo_test.exs (hunk 5 above)
Comment (hannahbit) [reply test/do_it/todo_test.exs:47]: Ah ok!

File: test/do_it/todo_test.exs (hunk 6 above)
Comment (hannahbit) [reply test/do_it/todo_test.exs:6]: Valid point 👍

File: test/do_it/todo_test.exs (hunk 7 above)
Comment (hannahbit) [reply test/do_it/todo_test.exs:1]: Ah, that is cool! I already did this for the Controller tests. Will do that 💯

File: lib/do_it/todo.ex (hunk 4 above)
Comment (tessi) [reply lib/do_it/todo.ex:18]: ah, I somehow missed the `validate_required` when writing this comment 🙈 

Forget what I said :)

File: lib/do_it/todo.ex (hunk 1 above)
Comment (tessi) [reply lib/do_it/todo.ex:17]: The "pass in the actually required struct" is especially useful if you have the `list` being around in your controller anyways (which is often the case in these kind of controllers in more complex apps) - this way you don't event need to write extra code :)

Comment (klappradla) [reply lib/do_it/todo.ex:17]: > Is this ok? Putting the :list_id as required?

In my opinion, that makes perfect sense @hannahbit 👍 

For your initial question about how to validate this: There is also [assoc_constraint/3](https://hexdocs.pm/ecto/Ecto.Changeset.html#assoc_constraint/3), maybe that fits your use case? Have a look at the example 👀

Comment (hannahbit) [reply lib/do_it/todo.ex:17]: @klappradla Thanks for your feedback! Yes I use the assoc_constraint already its on the next line :)

Comment (klappradla) [reply lib/do_it/todo.ex:17]: Haha, then sorry for the noise 🙈

What I was wondering though: does `assoc_constraint/3` work for you without having the `list_id` in the list of attributes you pass to `cast/2`? I'd personally intuitively pass every field which I want to use (and eventually validate further) into `cast/2` 🤔

Comment (hannahbit) [reply lib/do_it/todo.ex:17]: Yep. It does work. Would you it is better practice though to put `list_id` in `cast/2` ?

File: test/do_it/todo_test.exs [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+  use DoIt.DataCase, async: true
+  alias DoIt.Todo
+  alias DoIt.Repo
+
+  @long_description "This is a very long description. It is longer than onehundred characters. This is really very long. Who would give their todo list such a long description?"
+
+  test "description must be at least three characters long" do
+    changeset = Todo.create_changeset(%Todo{list_id: 1}, %{description: "In"})
```
Comment (klappradla) [inline test/do_it/todo_test.exs:9]: This may be cosmetic and slightly derailing, but it just came to my mind while reading this 😉 

Why to you pass a `Todo` struct with a `list_id` set? The list does not really have much to do with what's tested here. So I'd say it just generates noise and distracts from the contents of the test.

Furthermore, why is it necessary to pass a `Todo` struct to `create_changeset/2` at all? I imagine the use case for the function to be a scenario where you don't have a `Todo` item yet. Otherwise, you'd probably use an `update_changeset` function 🤔 

So your implementation of `create_changeset/2` could in fact also be a `create_changeset/1` and look roughly like that:
```elixir
def create_changeset(attrs)
  %__MODULE__
  |> cast([:description, :list_id…])
  |> validate ...
end
```

Doing so, all attributes (including the `list_id`) are passed in via the `attrs` and there is no mix between already set fields on the `Todo` struct, which I'd expect to not even exist yet when doing a "create" operation.

Anyways, just a thought 🤷

File: lib/do_it/todo.ex [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+  schema "todos" do
+    belongs_to :list, DoIt.List
+    field :checked, :boolean, default: false
+    field :description, :string
+
+    timestamps()
+  end
+
+  @doc false
+  def create_changeset(todo, attrs) do
```
Comment (klappradla) [inline lib/do_it/todo.ex:14]: Expanding on my [comment in the tests](https://github.com/hannahbit/do_it/pull/10/files#r650751413): the way I ready this function, it can currently do both, change an existing `Todo` item **and** generate one from the ground up. Nitpicking, but I feel that the name could be slightly misleading then: I'd expect a function intended to build a changeset for a `create` operation to not even take an existing item.

If you however intend to use the changeset function for both, "update" and "create" operations, I'd suggest to also rename it to something more generic, e.g. just `changeset/2`.

Again, just something that came to my mind when reading this - only cosmetics.

File: test/do_it/todo_test.exs (hunk 8 above)
Comment (hannahbit) [reply test/do_it/todo_test.exs:9]: 🤔 My thinking was, that I would probably use build_assoc/3 in the controller before and like this already generate a `Todo` . Apparently there is so many ways to set this association and a lot of nuances to understand, to pick the ideal implementation 🙃

File: lib/do_it/todo.ex (hunk 9 above)
Comment (hannahbit) [reply lib/do_it/todo.ex:14]: No I think I just want to use it for #create. So I'll go with your suggestion from your comment in the tests. 👍

File: lib/do_it/todo.ex (hunk 1 above)
Comment (klappradla) [reply lib/do_it/todo.ex:17]: > Yep. It does work. Would you say it is better practice though to put `list_id` in `cast/2` ?

Yes.

File: test/do_it/todo_test.exs (hunk 8 above)
Comment (klappradla) [reply test/do_it/todo_test.exs:9]: > Apparently there is so many ways to set this association and a lot of nuances to understand, to pick the ideal implementation 🙃

I know 🙈 

But I think a few more general ideas can always help shipping around in that swamp 🚣:
- try to compose things of manageable, self contained small blocks 🧱
- separate concerns and let the blocks only know the absolute minimum they need to perform their task 🐥

With regards to these, I'd say the less the controller knows about the underlying details of the `Todo` schema, the better. So if it just passes all the parameters to the changeset function, the details of fiddling these attributes together to an actually valid `Todo` item stay in the `Todo` model and the controller can concentrate on its own core tasks: accept data, pass it to business logic and render out the result.

Comment (hannahbit) [reply test/do_it/todo_test.exs:9]: Thanks! I will keep a screenshot of that and come back to it whenever I am in doubt 😉
